surfman/src/platform/egl/surface/ohos_surface.rs

```diff
@@ -0,0 +1,280 @@
+// surfman/surfman/src/platform/egl/ohos_surface.rs
+//
+//! Surface management for OpenHarmony OS using EGL.
+
+use std::marker::PhantomData;
+use std::os::raw::c_void;
+use std::ptr;
+
+use euclid::default::Size2D;
+use log::info;
+
+use crate::egl;
+use crate::egl::types::EGLSurface;
+use crate::gl;
+use crate::gl::types::{GLenum, GLuint};
+use crate::platform::generic::egl::device::EGL_FUNCTIONS;
+use crate::platform::generic::egl::ffi::EGL_EXTENSION_FUNCTIONS;
+use crate::platform::generic::egl::ffi::EGL_NO_IMAGE_KHR;
+use crate::{Error, SurfaceAccess, SurfaceID, SurfaceInfo, SurfaceType};
+
+use super::super::context::{Context, GL_FUNCTIONS};
+use super::super::device::Device;
+use super::super::ohos_ffi::{
+    NativeWindowOperation, OHNativeWindow, OH_NativeWindow_NativeWindowHandleOpt,
+};
+use super::{Surface, SurfaceTexture};
+
+const SURFACE_GL_TEXTURE_TARGET: GLenum = gl::TEXTURE_2D;
+
+pub(crate) enum SurfaceObjects {
+    Window { egl_surface: EGLSurface },
+}
+
+/// An OHOS native window.
+pub struct NativeWidget {
+    pub(crate) native_window: *mut OHNativeWindow,
+}
+
+impl Device {
+    /// Creates either a generic or a widget surface, depending on the supplied surface type.
+    ///
+    /// Only the given context may ever render to the surface, but generic surfaces can be wrapped
+    /// up in a `SurfaceTexture` for reading by other contexts.
+    pub fn create_surface(
+        &mut self,
+        context: &Context,
+        _: SurfaceAccess,
+        surface_type: SurfaceType<NativeWidget>,
+    ) -> Result<Surface, Error> {
+        info!("Device create_surface with Context");
+        match surface_type {
+            SurfaceType::Generic { size } => self.create_generic_surface(context, &size),
+            SurfaceType::Widget { native_widget } => unsafe {
+                self.create_window_surface(context, native_widget)
+            },
+        }
+    }
+
+    fn create_generic_surface(
+        &mut self,
+        _context: &Context,
+        _size: &Size2D<i32>,
+    ) -> Result<Surface, Error> {
+        Err(Error::Unimplemented)
+    }
+
+    unsafe fn create_window_surface(
+        &mut self,
+        context: &Context,
+        native_widget: NativeWidget,
+    ) -> Result<Surface, Error> {
+        let mut height: i32 = 0;
+        let mut width: i32 = 0;
+        // Safety: `OH_NativeWindow_NativeWindowHandleOpt` takes two output i32 pointers as
+        // variable arguments when called with `GET_BUFFER_GEOMETRY`. See the OHNativeWindow
+        // documentation for details:
+        // https://gitee.com/openharmony/docs/blob/master/en/application-dev/reference/apis-arkgraphics2d/_native_window.md
```

Comment:
Perhaps this should move to a rustdoc safety section?

Reply:
This comment is intended for reviewers of the unsafe block, since `OH_NativeWindow_NativeWindowHandleOpt` takes different arguments depending on the value of second parameter, and thus needs to be manually cross-referenced with the documentation during review.
Callers of this function don't need to know about `OH_NativeWindow_NativeWindowHandleOpt`. 

To be honest I'm not really sure what the Safety preconditions of `create_window_surface()` should be (and if it needs to be unsafe). But I'm pretty sure the Safety preconditions should be the same as the android version (which isn't documented either, though)

Reply:
I added a Safety section to the rustdoc of the function declaration in `ohos_ffi.rs`

Reply:
I agree with your original statement and IMO the original placement isn't an issue. I've just seen that clippy often complains if an unsafe function does not have a "# Safety" section. It just seems that since you have already gone through the trouble of writing a nice comment here, we can use it to make clippy happy as well. :D